server/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportGetAllocationStatsAction.java

```diff
@@ -185,4 +213,42 @@ public DiskThresholdSettings getDiskThresholdSettings() {
             return diskThresholdSettings;
         }
     }
+
+    private record CachedAllocationStats(Map<String, NodeAllocationStats> stats, long timestampMsecs) {}
+
+    private static class AllocationStatsCache {
+        private volatile long maxAgeMsecs;
+        private final AtomicReference<CachedAllocationStats> cachedStats;
+
+        AllocationStatsCache(long maxAgeMsecs) {
+            this.maxAgeMsecs = maxAgeMsecs;
+            this.cachedStats = new AtomicReference<>();
+        }
+
+        void setMaxAgeMsecs(long maxAgeMsecs) {
+            this.maxAgeMsecs = maxAgeMsecs;
+        }
+
+        Map<String, NodeAllocationStats> get() {
+
+            if (maxAgeMsecs == CACHE_DISABLED) {
+                return null;
+            }
+
+            final var stats = cachedStats.get();
+
+            if (stats == null || System.currentTimeMillis() - stats.timestampMsecs > maxAgeMsecs) {
```

Comment:
I would suggest using `org.elasticsearch.threadpool.ThreadPool#relativeTimeInMillis` rather than `System.currentTimeMillis` because (a) that way we can inject a different clock in tests more easily and (b) `System.currentTimeMillis` is not monotonic and if it jumps backwards then we might not invalidate the cache for a long time.